File: sdk/go/memory/memory.go
package memory

import (
	"crypto/sha256"
	"encoding/hex"
	"sync"

	"github.com/fhenix/fhe-agent-shield-go/fhe"
)

// Snapshot represents a memory snapshot
type Snapshot struct {
	ID        string
	AgentID   string
	Context   []string
	Timestamp uint64
}

// Memory manages FHE-encrypted agent memory
type Memory struct {
	client    *fhe.Client
	contracts fhe.ContractAddresses
	context   []string
	snapshots map[string]Snapshot
	mu        sync.RWMutex
}

// New creates a new memory manager
func New(client *fhe.Client, contracts fhe.ContractAddresses) *Memory {
	return &Memory{
		client:    client,
		contracts: contracts,
		context:   make([]string, 0),
		snapshots: make(map[string]Snapshot),
	}
}

// AppendContext adds encrypted context to memory
func (m *Memory) AppendContext(data string) (string, error) {
	m.mu.Lock()
	defer m.mu.Unlock()

	m.context = append(m.context, data)

	// In production, this would call the AgentMemory contract
	// Return mock tx hash
	txData := "tx_" + data
	hexStr := hex.EncodeToString([]byte(txData))
	if len(hexStr) > 40 {
		hexStr = hexStr[:40]
	}
	return "0x" + hexStr, nil
}

// GetContext retrieves recent memory context
func (m *Memory) GetContext(limit int) ([]string, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	if limit <= 0 || limit > len(m.context) {
		limit = len(m.context)
	}

	result := make([]string, limit)
	copy(result, m.context[len(m.context)-limit:])

	return result, nil
}

// ClearMemory clears all memory context
func (m *Memory) ClearMemory() (string, error) {
	m.mu.Lock()
	defer m.mu.Unlock()

	// Reset context slice to empty but keep capacity
	m.context = m.context[:0]

	// In production, this would call the AgentMemory contract
	// Return mock tx hash
	hexStr := hex.EncodeToString([]byte("clear"))
	if len(hexStr) > 40 {
		hexStr = hexStr[:40]
	}
	return "0x" + hexStr, nil
}

// CreateSnapshot creates a snapshot of current memory state
func (m *Memory) CreateSnapshot() (string, error) {
	m.mu.Lock()
	defer m.mu.Unlock()

	// Generate snapshot ID
	hash := sha256.Sum256([]byte(m.client.Address() + string(rune(len(m.snapshots)))))
	id := hex.EncodeToString(hash[:16])

	// Copy current context to snapshot
	ctxCopy := make([]string, len(m.context))
	copy(ctxCopy, m.context)

	m.snapshots[id] = Snapshot{
		ID:        id,
		AgentID:   m.client.Address(),
		Context:   ctxCopy,
		Timestamp: uint64(0), // Would be block timestamp in production
	}

	return id, nil
}

// RestoreSnapshot restores memory from a snapshot
func (m *Memory) RestoreSnapshot(snapshotID string) (string, error) {
	m.mu.Lock()
	defer m.mu.Unlock()

	snapshot, ok := m.snapshots[snapshotID]
	if !ok {
		return "", nil // Snapshot not found
	}

	m.context = make([]string, len(snapshot.Context))
	copy(m.context, snapshot.Context)

	return "0x" + hex.EncodeToString([]byte("restore_" + snapshotID))[:40], nil
}

// ListSnapshots returns all snapshots for the agent
func (m *Memory) ListSnapshots() ([]Snapshot, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	snapshots := make([]Snapshot, 0, len(m.snapshots))
	for _, s := range m.snapshots {
		if s.AgentID == m.client.Address() {
			snapshots = append(snapshots, s)
		}
	}

	return snapshots, nil
}

File: sdk/elixir/lib/memory.ex
defmodule FHEAgentShield.Memory do
  @moduledoc """
  FHE-encrypted agent memory management.

  Memory is stored on-chain using the AgentMemory contract.
  Snapshots allow state restoration.
  """

  alias FHEAgentShield.Client

  @doc """
  Append encrypted context to agent memory.
  """
  @spec append_context(Client.t(), String.t(), keyword()) ::
          {:ok, String.t()} | {:error, term()}
  def append_context(_client, encrypted_data, _opts \\ []) do
    # In production:
    # 1. Encrypt data using FHE
    # 2. Build transaction to AgentMemory.appendContext
    # 3. Send transaction
    # 4. Return tx hash
    {:ok, "0x..."}
  end

  @doc """
  Get recent memory context.
  """
  @spec get_context(Client.t(), keyword()) ::
          {:ok, [String.t()]} | {:error, term()}
  def get_context(_client, opts \\ []) do
    session_id = Keyword.get(opts, :session_id, "default")
    limit = Keyword.get(opts, :limit, 10)

    # In production:
    # 1. Query AgentMemory contract
    # 2. Decrypt results
    # 3. Return plaintext context
    {:ok, []}
  end

  @doc """
  Clear agent memory.
  """
  @spec clear_memory(Client.t()) :: {:ok, String.t()} | {:error, term()}
  def clear_memory(_client) do
    {:ok, "0x..."}
  end

  @doc """
  Create a memory snapshot.
  """
  @spec create_snapshot(Client.t()) :: {:ok, String.t()} | {:error, term()}
  def create_snapshot(_client) do
    # In production:
    # 1. Call AgentMemory.createSnapshot()
    # 2. Returns snapshot ID (address derived from hash)
    snapshot_id = "snapshot_#{:rand.uniform(999_999)}"
    {:ok, snapshot_id}
  end

  @doc """
  Restore from a memory snapshot.
  """
  @spec restore_snapshot(Client.t(), String.t()) ::
          {:ok, String.t()} | {:error, term()}
  def restore_snapshot(_client, snapshot_id) do
    # In production:
    # 1. Call AgentMemory.restoreFromSnapshot(snapshot_id)
    # 2. Returns tx hash
    {:ok, "0x..."}
  end

  @doc """
  List snapshots for the agent.
  """
  @spec list_snapshots(Client.t()) :: {:ok, [map()]} | {:error, term()}
  def list_snapshots(_client) do
    {:ok, []}
  end
end
